Allow rendering a text preview using custom fonts

jvillafanez opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the oCIS component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next version of oCIS.

Please set the following labels:

- Set label "Status:Needs-Review" for review or "Status:In-Progress" if the PR still has open tasks.
- Assignment: assign to self
- Reviewers: pick at least one
-->

## Description
Allow usage of custom fonts for the text previews so non-latin chars (or chars not supported by the "goregular" font) can be shown

## Related Issue
https://github.com/owncloud/ocis/issues/2620

## Motivation and Context
Files containing non-latin can have a proper preview

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- test environment:
- test case 1:
- test case 2:
- ...

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in owncloud.github.io/ -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

### Expected usage

Start the thumbnails service using the `THUMBNAILS_TXT_FONTMAP_FILE` pointing to a json file, for example `THUMBNAILS_TXT_FONTMAP_FILE=/myfonts/fontMap.json`. Note that all the fonts referred by the font map file must be downloaded manually and be present in the FS.

The json file must contain a json object with the following info:
* A "fontMap" key, which contains a map of script => font_file
* A "defaultFont" key, which contains the default font to be used in case the target script isn't found in the fontMap

An example of font map file can be the following
```
{
        "fontMap": {
                "Arabic": "NotoSansArabic-Regular.ttf",
                "Devanagari": "NotoSansDevanagari-Regular.ttf",
                "Latin": "NotoSans-Regular.ttf",
                "Hangul": "NotoSansMonoCJKkr-Regular.otf",
                "Han": "NotoSansMonoCJKsc-Regular.otf",
                "Hiragana": "NotoSansMonoCJKjp-Regular.otf",
                "Katakana": "NotoSansMonoCJKjp-Regular.otf"
        },
        "defaultFont": "NotoSans-Regular.ttf"
}
```
A full list of available scripts can be found in https://cs.opensource.google/go/go/+/refs/tags/go1.17.2:src/unicode/tables.go;l=3655

In case the target script isn't mapped, the default font will be used. The character of that script will likely appear as a white box instead of the expected character.

The font map contains paths to the font files. The paths can be absolute or relative as shown below:
```
{
        "fontMap": {
                "Arabic": "NotoSansArabic-Regular.ttf",
                "Devanagari": "NotoSansDevanagari-Regular.ttf",
                "Latin": "/customFonts/test001.ttf",
                "Hangul": "NotoSansMonoCJKkr-Regular.otf",
                "Han": "NotoSansMonoCJKsc-Regular.otf",
                "Hiragana": "NotoSansMonoCJKjp-Regular.otf",
                "Katakana": "NotoSansMonoCJKjp-Regular.otf"
        },
        "defaultFont": "/customFonts/test001.ttf"
}
```
If the path is absolute, the font will be loaded directly from that path.
If the path is relative, it will be relative to the directory where the font map file is placed. For example, if we assume that the "/fontMaps/fontMap1.json" file contains the content above, the "NotoSans*" paths will refer to "/fontMaps/NotoSans*"
This allows the usage of self-contained packages of fonts that are ready to be used. The thumbnails service would just need to point to the font map file contained in the package.

### Known issues

* Arabic (and likely any right-to-left) script won't be rendered properly. Glyphs will be printed left-to-right
* Glyphs that might require repositioning or that are a combination of 2 or more glyphs, won't be rendered properly. This has been noticed with the Devanagari script such as `विश्वस्त हुनुहोस् तपाइँको`. Glyphs will be printed separately one after another.
* Punctuation symbols might not be rendered properly in some circumstances. This usually involves using multiple scripts such as in `附加到电子邮件的未版本化文档以及安全性有问题的公共云中的影子 IT。`. In the example, the last character is used in chinese or japanese languages, and latin fonts (such as NotoSans-Regular.ttf) doesn't contain that char. The problem is that the char is wrongly interpreted as latin because it's preceded by a latin char.
* Text file MUST be encoded in UTF-8. Text encoded in any other format will show random chars.

### To Do List
* Include tests
* Check possible concurrency problems (docs say that the `font.Face` isn't thread-safe)

ownclouders commented:
<details><summary>:boom: Acceptance tests <strong>Web-Tests-ocis-ocis-storage-1</strong> failed. The build is cancelled...</summary>

jvillafanez commented:
We'll leave the known issues aside.

Both first and second issues are too complex to tackle on our own just for a text preview. It involves a text shaping mechanism which isn't present in the "golang.org/x/image" package, and I didn't found alternative packages (maybe https://github.com/benoitkugler/textlayout).
Note that this goes beyond the initial thought of the problem. It isn't "just" draw the chars from right to left, or reposition a char, but substituting a combination of glyphs for a different one.

ownclouders commented:
<details><summary>:boom: Acceptance tests <strong>accountsUITests</strong> failed. The build is cancelled...</summary>

ownclouders commented:
<details><summary>:boom: Acceptance tests <strong>Web-Tests-ocis-ocis-storage-1</strong> failed. The build is cancelled...</summary>

ownclouders commented:
:boom: Acceptance test [<strong>localApiTests-apiArchiver-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/7773/21/1) failed. Further test are cancelled...
<!-- id: pr-2723 -->

C0rby reviewed (approved):
LGTM

refs reviewed (approved):
Tried it locally and works just fine 👍 rebase and merge 👯

jvillafanez commented:
Rebased. Note that there have been conflicts and the code has been slightly changed:
```
juan@juan-VBox:~/src/ocis/ocis$ git rebase origin/master
First, rewinding head to replay your work on top of it...
Applying: Allow rendering a text preview using custom fonts
Using index info to reconstruct a base tree...
M	thumbnails/pkg/config/config.go
A	thumbnails/pkg/flagset/flagset.go
Falling back to patching base and 3-way merge...
CONFLICT (modify/delete): thumbnails/pkg/flagset/flagset.go deleted in HEAD and modified in Allow rendering a text preview using custom fonts. Version Allow rendering a text preview using custom fonts of thumbnails/pkg/flagset/flagset.go left in tree.
Auto-merging thumbnails/pkg/config/config.go
CONFLICT (content): Merge conflict in thumbnails/pkg/config/config.go
error: Failed to merge in the changes.
Patch failed at 0001 Allow rendering a text preview using custom fonts
hint: Use 'git am --show-current-patch' to see the failed patch
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```
* thumbnails/pkg/flagset/flagset.go -> deleted
* thumbnails/pkg/config/config.go -> adjusted
* thumbnails/pkg/config/mappings.go -> adjusted

I can recheck tomorrow to ensure it still works after the rebase.

jvillafanez commented:
Looks good to me. Ready to merge.
